File: Example Code/Java/com/sensorcloud/Requests.java
/*
 * Copyright 2012 MicroStrain Inc. All Rights Reserved.
 *
 * Distributed under the Simplified BSD License.
 * See file license.txt
 *
 */

package com.sensorcloud;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.ProtocolException;
import java.net.URL;
import java.util.Hashtable;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;

public class Requests
{
	public static class KeyValueStore 
	{
		private Hashtable<String, String> map;
		
		public KeyValueStore()
		{
			map = new Hashtable<String, String>();
		}
		
		@SuppressWarnings("unchecked")
		public KeyValueStore(Hashtable<String, String> hashtable)
		{
			map = (Hashtable<String, String>)hashtable.clone();
		}
		
		public void put(String key, String value)
		{
			map.put(key,  value);
		}
		
		public String get(String key)
		{
			return map.get(key);
		}
		
		public Hashtable<String, String> getMap()
		{
			return map;
		}
		
		public boolean isEmpty()
		{
			return map.isEmpty();
		}
		
		public int size()
		{
			return map.size();
		}
		
		public Set<Entry<String,String>> entrySet()
		{
			return map.entrySet();
		}
		
		
		@Override
		public boolean equals (Object obj)
		{
			if (!(obj instanceof KeyValueStore)) { return false; }
			return map.equals(((KeyValueStore)obj).map);
		}
		
		
		@Override
		public int hashCode()
		{
			return  map.hashCode();
		}	
	}
	
	public static class Params extends KeyValueStore
	{
		
		public String queryString()
		{
			StringBuilder s = new StringBuilder();
			for (Map.Entry<String, String> entry : entrySet())
			{
				s.append(s.length()==0?"?":"&");
				s.append(urlEncode(entry.getKey()));
				s.append("=");
				s.append(urlEncode(entry.getValue()));
			}
			
			return s.toString();
		}
		
		private String urlEncode(String s)
		{
			//TODO: need to escape the strings that we are placing on the url path
			//we should find an existing class that takes care of this instead of rolling our own
			return s;
		}
	}
	
	public static class Headers extends KeyValueStore{}
	
	public static class RequestOptions
	{
		private Params params;
		private Headers headers;
		private  byte[] payload;
	
		public RequestOptions()
		{
			this.params = new Params();
			this.headers = new Headers();
			this.payload = null;
		}
		
		public RequestOptions(Params params, Headers headers, byte[] data)
		{
			assert(params != null);
			assert(headers != null);
			
			this.params = params;
			this.headers = headers;
			this.payload = data;
		}
		
		public void addParam(String name, String value)
		{
			params.put(name, value);
		}
		
		public void addHeader(String name, String value)
		{
			headers.put(name, value);
		}
		
		public void setPayload(ByteArrayOutputStream payloadStream)
		{
			this.payload = payloadStream.toByteArray();
		}
		
		public void setPayload(String payload)
		{
			this.payload = payload.getBytes();		
		}
		
		public void setPayload(byte[] payload)
		{
			this.payload = payload;
		}
		
		public Params getParams(){return params;}
		public Headers getHeaders(){return headers;}
		
		public byte[] getData(){return payload;}
	}
	
	
	public static class RequestMaker
	{
		String url;
		RequestOptions options;
		
		public RequestMaker(String url)
		{
			this.url = url;
			this.options = new RequestOptions();
		}
		
		public RequestMaker header(String name, String value)
		{
			this.options.addHeader(name, value);
			return this;
		}
		
		public RequestMaker contentType(String contentType)
		{
			return header("Content-Type", contentType);
		}
		
		public RequestMaker accept(String acceptType)
		{
			return header("Accept", acceptType);
		}
		
		
		public RequestMaker param(String name, String value)
		{
			this.options.addParam(name, value);
			return this;
		}
		
		public RequestMaker data(ByteArrayOutputStream payload)
		{
			this.options.setPayload(payload);
			return this;
		}
		
		public Request get() throws IOException 
		{
			return new Request("GET", this.url, this.options);
		}
		
		public Request put() throws IOException 
		{
			return new Request("PUT", this.url, this.options);
		}
		
		public Request post() throws IOException 
		{
			return new Request("POST", this.url, this.options);
		}
		
		public Request delete() throws IOException 
		{
			return new Request("DELETE", this.url, this.options);
		}
	}
	
	
	public static class Request
	{
		private String method;
		private String url;
		private RequestOptions options;
		
		private int responseCode;
		private String responseMessage;
			
		private InputStream responseStream; 
		
		public Request(String method, String url, Params params, Headers headers, byte[] data) throws IOException
		{
			assert(params != null);
			assert(headers != null);
			
			this.method = method;
			this.url = url;
			this.options = new RequestOptions(params, headers, data);
			
			this.responseCode = 0;
			this.responseMessage = "";
			this.responseStream = null;
			
			doRequest();
		}
		
		public Request(String method,String url, RequestOptions options) throws IOException
		{
			this.method = method;
			this.url = url;
			this.options = options;
			
			this.responseCode = 0;
			this.responseMessage = "";
			this.responseStream = null;
			
			doRequest();
		}
		
		public int getResponseCode()
		{
			return this.responseCode;
		}
		
		public String getResponseMessage()
		{
			return this.responseMessage;
		}
		
		public InputStream getResponseStream()
		{
			return this.responseStream;
		}
		
		public String getResponseData() throws IOException
		{
			return new String(this.getRawResponseData());
		}
		
		public byte[] getRawResponseData() throws IOException
		{
			if(this.responseStream == null)
			{
				return null;
			}
			
			ByteArrayOutputStream os = new ByteArrayOutputStream();
				
			byte [] buf = new byte [1024];		
			int len;
			while((len = this.responseStream.read(buf)) > 0)
			{
				os.write(buf, 0, len);
			}		
			os.flush();
						
			return os.toByteArray();	
		}
		
		
		private void doRequest() throws IOException
		{
			//possible exception - UnknownHostException, MalformedURLException
			
			
			//build url
			String q = options.getParams().queryString();
			String strUrl = this.url + q;
			
			URL url = new URL(strUrl);
			
			HttpURLConnection connection = (HttpURLConnection) url.openConnection();
			
			try
			{
				connection.setRequestMethod(this.method);
			}
			catch (ProtocolException e)
			{
				//Should never happen. user never sets method directly.
				assert(false);
			}
			
			//add all the headers to the request
			for (Map.Entry<String, String> entry : options.getHeaders().entrySet())
			{
				connection.setRequestProperty(entry.getKey(), entry.getValue());
			}
			
			//send data to the server
			if(options.getData() != null)
			{
				connection.setDoOutput(true);
				OutputStream os = connection.getOutputStream();
				os.write(options.getData());
			}	
		
			this.responseCode = connection.getResponseCode();
			this.responseMessage = connection.getResponseMessage();
			
			// depending on the status code, the response will sometimes be in the error stream
			 this.responseStream = connection.getErrorStream();
			if(this.responseStream == null)
			{
				//call getInputStream will raise an exception with a 404,
				 this.responseStream = connection.getInputStream();
			}
		}
	}
	
	public static RequestMaker url(String url)
	{
		return new RequestMaker(url);
	}
}

File: SDK/CSharp/Requests.cs
/*
 * Copyright 2012 MicroStrain Inc. All Rights Reserved.
 *
 * Distributed under the Simplified BSD License.
 * See file license.txt
 *
 */

using System;
using System.Collections.Generic;
using System.Diagnostics;

using System.Web;
using System.Net;
using System.IO;


namespace SensorCloud
{
	/// <summary>
	/// Wrapper around HttpWebRequest to simplify the creation of http requests.
	/// It was specifically designed for interacting with web services.
	///
	/// The intention is for this class to be a generic http request, but it was built
	/// explicitly for sensorcloud, so only the options that are required for SensorCloud
	/// have been implemented. 
	/// </summary>
	public class Requests : IRequests
	{		
		/// <summary>
		///  Dictionary used exclusively for url parameters
		/// </summary>
		public class Params : Dictionary<String, String>
		{
			public String QueryString()
			{
				String s = "";
				foreach (var pair in this)
				{
					s += s.Length == 0 ? "?" : "&";
					s += HttpUtility.UrlEncode(pair.Key) + "=" + HttpUtility.UrlEncode(pair.Value);
				}
				return s;
			}
		}


		/// <summary>
		/// Dictionary used exclusively for http headers
		/// </summary>
		public class Headers : Dictionary<String, String> { }


		/// <summary>
		/// RequestOptions stores all the possible variables required to make an http request
		/// </summary>
		public class RequestOptions
		{
			private Params _queryParams;
			private Headers _headers;
			private byte[] _requestBody;

			public Headers Headers { get { return this._headers; } }
			public String QueryString { get { return this._queryParams.QueryString(); } }
			public byte[] RequestBody { get { return _requestBody; } }

			public RequestOptions()
			{
				this._queryParams = new Params();
				this._headers = new Headers();
				this._requestBody = null;
			}


			/// <summary>
			/// Add a query string param to the request options
			/// </summary>
			/// <param name="name"></param>
			/// <param name="value"></param>
			public void AddParam(String name, String value)
			{
				_queryParams.Add(name, value);
			}


			/// <summary>
			/// Add an http header to the request options
			/// </summary>
			/// <param name="name"></param>
			/// <param name="value"></param>
			public void AddHeader(String name, String value)
			{
				_headers.Add(name, value);
			}

			/*
			public void SetPayload(Stream requestBodyStream)
			{
				this._requestBody = requestBodyStream;
			}
		
			public void SetPayload(String requestBody)
			{
				this._requestBody = requestBody;		
			}
		*/

			/// <summary>
			/// Adds data that is sent as the body of a request
			/// </summary>
			/// <param name="requestBody"></param>
			public void SetRequestBody(byte[] requestBody)
			{
				this._requestBody = requestBody;
			}
		}



		/// <summary>
		/// An http request has a wide range of possible options that can be specified.
		/// RequestBuilder parameterizes each possible options as a method returning a RequestBuilder option
		/// allowing the chaining of request options.  The chaining allows us to eliminate the need for having an overloaded
		/// method that can handle any combination of desired options, and also results in a very readable self documented
		/// request.
		/// </summary>
		public class RequestBuilder : IRequestBuilder
		{
			private String _url;
			private RequestOptions _options;

			public RequestBuilder(String url)
			{
				this._url = url;
				this._options = new RequestOptions();
			}

			/// <summary>
			/// Add an http header to the request
			/// </summary>
			/// <param name="name"></param>
			/// <param name="value"></param>
			/// <returns></returns>
			public IRequestBuilder Header(String name, String value)
			{
				this._options.AddHeader(name, value);
				return this;
			}

			/// <summary>
			/// Set the content Type for the request's messageBody.
			/// This is a convenience method for setting the HTTP header "Content-Type".
			/// </summary>
			/// <param name="contentType"></param>
			/// <returns></returns>
			public IRequestBuilder ContentType(String contentType)
			{
				return Header("Content-Type", contentType);
			}

			/// <summary>
			/// specifies the content-type or types that we can handle in the response.
			/// This is a convenience method for seeting the HTTP header "Accept".
			/// </summary>
			/// <param name="acceptType"></param>
			/// <returns></returns>
			public IRequestBuilder Accept(String acceptType)
			{
				return Header("Accept", acceptType);
			}


			/// <summary>
			/// Add a query parameter to the request
			/// </summary>
			/// <param name="name"></param>
			/// <param name="value"></param>
			/// <returns></returns>
			public IRequestBuilder Param(String name, String value)
			{
				this._options.AddParam(name, value);
				return this;
			}

			/// <summary>
			/// Add a query parameter to the request for any object.  Uses the object's ToString to convert it to a string
			/// </summary>
			/// <param name="name"></param>
			/// <param name="value"></param>
			/// <returns></returns>
			public IRequestBuilder Param(String name, Object value)
			{
				return Param(name, value.ToString());
			}

			/// <summary>
			/// Add a message body to the request.
			/// </summary>
			/// <param name="requestBody"></param>
			/// <returns></returns>
			public IRequestBuilder Data(byte[] requestBody)
			{
				this._options.SetRequestBody(requestBody);
				return this;
			}

			/// <summary>
			/// Invoke the request that has been built using the GET verb
			/// </summary>
			/// <returns></returns>
			public IRequest Get()
			{
				return DoRequest("GET", this._url, this._options);
			}

			/// <summary>
			/// Invoke the request that has been built using the PUT verb
			/// </summary>
			/// <returns></returns>
			public IRequest Put()
			{
				return DoRequest("PUT", this._url, this._options);
			}

			/// <summary>
			/// Invoke the request that has been built using the POST verb
			/// </summary>
			/// <returns></returns>
			public IRequest Post()
			{
				return DoRequest("POST", this._url, this._options);
			}

			/// <summary>
			/// Invoke the request that has been built using the DELETE verb
			/// </summary>
			/// <returns></returns>
			public IRequest Delete()
			{
				//return new Request("DELETE", this._url, this._options);
				return DoRequest("DELETE", this._url, this._options);
			}

			protected virtual IRequest DoRequest(string method, string url, RequestOptions options)
			{
				return new Request(method, url, options);
			}
		}

		public class Request : IRequest
		{
			protected String _method;
			protected String _url;
			protected RequestOptions _options;


			private byte[] responseData;

			public HttpStatusCode ResponseCode { get; private set; }
			public String ResponseMessage { get; private set; }
			public String ContentType { get; private set; }

			public String Text
			{
				get
				{
					if (responseData == null) { return ""; }
					var encoding = new System.Text.UTF8Encoding();
					return encoding.GetString(responseData);
				}
			}


			public byte[] Raw { get { return responseData; } }


			public Request(String method, String url, RequestOptions options)
			{
				this._method = method;
				this._url = url;
				this._options = options;

				this.ResponseCode = 0;
				this.ResponseMessage = "";
				this.ContentType = "";
				this.responseData = null;

				DoRequest();
			}


			private void DoRequest()
			{
				//build url
				String q = _options.QueryString;
				String strUrl = this._url + q;

				var request = (HttpWebRequest)WebRequest.Create(strUrl);
				request.Method = this._method;

				//add all the headers to the request
				foreach (KeyValuePair<String, String> pair in _options.Headers)
				{
					var key = pair.Key.ToLowerInvariant();
					switch (key)
					{
						case "accept": request.Accept = pair.Value; break;
						case "content-type": request.ContentType = pair.Value; break;
						case "expect": request.Expect = pair.Value; break;
						case "host": request.Host = pair.Value; break;
						case "referer": request.Referer = pair.Value; break;
						case "transfer-encoding": request.TransferEncoding = pair.Value; break;
						case "user-agent": request.UserAgent = pair.Value; break;

						//TODO: These require actual date objects to be set, but we don't currently use these headers
						//case "if-modified-since": request.IfModifiedSince = pair.Value; break;
						//case "date": request.Date:

						default: request.Headers.Add(pair.Key, pair.Value); break;
					}
				}

				//send data to the server
				if (_options.RequestBody != null)
				{
					using (var stream = request.GetRequestStream())
					{
						stream.Write(_options.RequestBody, 0, _options.RequestBody.Length);
						stream.Close();
					}
				}

				try
				{
					using (var response = (HttpWebResponse)request.GetResponse())
					{
						this.ResponseCode = response.StatusCode;
						this.ResponseMessage = response.StatusDescription;
						this.ContentType = response.ContentType;

						using (var responseStream = response.GetResponseStream())
						using (var ms = new MemoryStream())
						{
							responseStream.CopyTo(ms);
							this.responseData = ms.ToArray();
						}
					}
				}
				catch (WebException e)
				{
					if (e.Response is HttpWebResponse)
					{
						var response = (HttpWebResponse)e.Response;
						this.ResponseCode = response.StatusCode;
						this.ResponseMessage = response.StatusDescription;
						this.ContentType = response.ContentType;

						using (var responseStream = response.GetResponseStream())
						{
							using (var ms = new MemoryStream())
							{
								responseStream.CopyTo(ms);
								this.responseData = ms.ToArray();
							}
						}
					}
					else
					{
						throw;
					}
				}
			}
		}

	
		/// <summary>
		/// Initiate the begining of a request using the url.  A request builder is returned allowing the request to be customized before 
		/// it is executed.
		/// </summary>
		/// <param name="url"></param>
		/// <returns></returns>
		public IRequestBuilder url(String url)
		{
			return new RequestBuilder(url);
		}
	}
}